### PR TITLE
Added methods to get qos defined in XML Profile [9607]

### PR DESCRIPTION
**Pull request by imontesino**

### CODE REVIEW

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 1]
```diff
@@ -625,6 +625,20 @@ const PublisherQos& DomainParticipantImpl::get_default_publisher_qos() const
     return default_pub_qos_;
 }
 
+const PublisherQos DomainParticipantImpl::get_publisher_qos_from_profile(
+        const std::string& profile_name) const
+{
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:630]: Change the method signature as explained above (apply also to `get_subscriber_qos_from_profile` and `get_topic_qos_from_profile`.
Follow the same logic used in `DomainParticipantFactory::get_participant_qos_from_profile`

File: src/cpp/fastdds/domain/DomainParticipant.cpp [hunk 2]
```diff
@@ -275,6 +275,14 @@ ReturnCode_t DomainParticipant::get_default_publisher_qos(
     return ReturnCode_t::RETCODE_OK;
 }
 
+ReturnCode_t DomainParticipant::get_publisher_qos_from_profile(
+        const std::string& profile_name,
+        PublisherQos& qos) const
+{
+    qos = impl_->get_publisher_qos_from_profile(profile_name);
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/domain/DomainParticipant.cpp:282]:
```suggestion
    return impl_->get_publisher_qos_from_profile(profile_name, qos);
```

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
There are a couple of things that should be done for all the methods. I have not written a comment every time a change should be made.

The review is summed up in:
* Use RETCODE_BAD_PARAMETER when the profile is not found. Do not return RETCODE_OK always.
* Modify the Impl methods to use the same signature as the non-Impl methods.
```
ReturnCode_t get_[ENTITY]_qos_from_profile(
        const std::string& profile_name,
        [ENTITY]Qos& qos) const;
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2879/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/182/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2634/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2890/)

File: src/cpp/fastdds/domain/DomainParticipantFactory.cpp [hunk 3]
```diff
@@ -39,6 +39,7 @@ using eprosima::fastdds::dds::Log;
 
 using eprosima::fastrtps::rtps::RTPSDomain;
 using eprosima::fastrtps::rtps::RTPSParticipant;
+using eprosima::fastrtps::rtps::RTPSParticipantAttributes;
```
[the remark this answers is not in the document]
Comment (imontesino) [reply src/cpp/fastdds/domain/DomainParticipantFactory.cpp:42]: No, it was from an earlier version of this functionality.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2882/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/184/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2637/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2893/)

File: include/fastdds/dds/domain/DomainParticipant.hpp [hunk 4]
```diff
@@ -372,6 +372,16 @@ class DomainParticipant : public Entity
     RTPS_DllAPI ReturnCode_t get_default_publisher_qos(
             PublisherQos& qos) const;
 
+    /**
+     * Fills the PublisherQos with the values of the XML profile.
+     * @param profile_name Publisher profile name.
+     * @param qos PublisherQos object where the qos is returned.
+     * @return RETCODE_OK
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply include/fastdds/dds/domain/DomainParticipant.hpp:379]: Please, update Doxygen documentation for all the previous methods explaining the return RETCODEs.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Please, update Doxygen documentation taking into account that the methods could also return RETCODE_BAD_PARAMETER

**Review by JLBuenoLopez-eProsima — DISMISSED** [review]
LGTM. Please add tests to this new feature and update the documentation.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2897/)
* Linux aarch64Build Status
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2907/)

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 5]
```diff
@@ -191,6 +192,29 @@ TEST(ParticipantTests, CreateDomainParticipantWithProfile)
...
 
+TEST(ParticipantTests, GetParticipantProfileQos)
+{
+    DomainParticipantFactory::get_instance()->load_XML_profiles_file("test_xml_profiles.xml");
+    DomainParticipantQos qos;
+    ASSERT_EQ(
+        DomainParticipantFactory::get_instance()->get_participant_qos_from_profile("test_participant_profile", qos),
+        ReturnCode_t::RETCODE_OK);
+
+    // Extract ParticipantQos form profile
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:203]:
```suggestion
    // Extract ParticipantQos from profile
```

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 6]
```diff
@@ -191,6 +192,29 @@ TEST(ParticipantTests, CreateDomainParticipantWithProfile)
...
+    DomainParticipant* participant =
+            DomainParticipantFactory::get_instance()->create_participant(0, qos);
+    ASSERT_NE(participant, nullptr);
+
+    check_participant_with_profile(participant, "test_participant_profile");
+
+    // Clean up
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:210]: This comment should be after this instruction. Here something like
`// Test return when a non-existent profile is used`
should be.

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 7]
```diff
@@ -538,6 +562,30 @@ void check_subscriber_with_profile (
...
+            DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
+
+    // Extract qos from profile
+    SubscriberQos qos;
+    ASSERT_EQ(
+        participant->get_subscriber_qos_from_profile("test_subscriber_profile", qos),
+        ReturnCode_t::RETCODE_OK);
+
+    Subscriber* subscriber = participant->create_subscriber(qos);
+
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:578]: You should check that the subscriber has been correctly created

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 8]
```diff
@@ -559,6 +607,31 @@ TEST(ParticipantTests, CreateSubscriberWithProfile)
...
+            DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
+
+    // Extract qos from profile
+    PublisherQos qos;
+    ASSERT_EQ(
+        participant->get_publisher_qos_from_profile("test_publisher_profile", qos),
+        ReturnCode_t::RETCODE_OK);
+
+    Publisher* publisher = participant->create_publisher(qos);
+
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:623]: Same as above: check that it has been correctly created

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 9]
```diff
@@ -559,6 +607,31 @@ TEST(ParticipantTests, CreateSubscriberWithProfile)
...
+    check_publisher_with_profile(publisher, "test_publisher_profile");
+
+    // Clean up
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:626]: This comment should be after this instruction

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 10]
```diff
@@ -669,6 +742,47 @@ TEST(ParticipantTests, ChangePSMDefaultTopicQos)
...
+    TypeSupport type(new TopicDataTypeMock());
+    type.register_type(participant);
+
+    // Extract qos from profile
+    TopicQos qos;
+    ASSERT_EQ(
+        participant->get_topic_qos_from_profile("test_topic_profile", qos),
+        ReturnCode_t::RETCODE_OK);
+
+    Topic* topic = participant->create_topic("footopic", type.get_type_name(), qos);
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:774]: Check that it has been created correctly

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 11]
```diff
@@ -669,6 +742,47 @@ TEST(ParticipantTests, ChangePSMDefaultTopicQos)
...
+
+    check_topic_with_profile(topic, "test_topic_profile");
+
+    // Clean up
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:778]: Move comment after following instruction

File: test/unittest/dds/publisher/PublisherTests.cpp [hunk 12]
```diff
@@ -296,6 +296,38 @@ TEST(PublisherTests, CreateDataWriterWithProfile)
     ASSERT_EQ(DomainParticipantFactory::get_instance()->delete_participant(participant), ReturnCode_t::RETCODE_OK);
 }
 
+TEST(PublisherTests, GetDataWriterProfileQos)
+{
+    DomainParticipantFactory::get_instance()->load_XML_profiles_file("test_xml_profiles.xml");
+    DomainParticipant* participant =
+            DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
+    Publisher* publisher = participant->create_publisher(PUBLISHER_QOS_DEFAULT);
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/publisher/PublisherTests.cpp:304]: Check all entities are correctly created as it is done in other tests

File: test/unittest/dds/subscriber/SubscriberTests.cpp [hunk 13]
```diff
@@ -330,6 +332,38 @@ TEST(SubscriberTests, CreateDataReaderWithProfile)
     ASSERT_EQ(DomainParticipantFactory::get_instance()->delete_participant(participant), ReturnCode_t::RETCODE_OK);
 }
 
+TEST(SubscriberTests, GetDataReaderProfileQos)
+{
+    DomainParticipantFactory::get_instance()->load_XML_profiles_file("test_xml_profiles.xml");
+    DomainParticipant* participant =
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/subscriber/SubscriberTests.cpp:338]: Same comments as in other tests

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 14]
```diff
@@ -547,7 +595,7 @@ TEST(ParticipantTests, CreateSubscriberWithProfile)
     //subscriber using the default profile
     Subscriber* default_subscriber = participant->create_subscriber(SUBSCRIBER_QOS_DEFAULT);
     ASSERT_NE(default_subscriber, nullptr);
-    check_subscriber_with_profile(default_subscriber, "test_default_subscriber_profile");
+    check_subscriber_with_profile(default_subscriber, "test_subscriber_profile");
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:598]: Why have you change this test?

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
Consider the following points:
* Check after creation of each entity that it has been correctly created (as it is done in other tests)
* Maybe you can use EXPECT_* instead of ASSERT_*. EXPECT makes the test fail too but it continues the test until the end. Only use EXPECT when the test could not keep running (i.e. Use ASSERT to check the creation of the entities, because if the entity is not created the test could not keep going)

File: test/unittest/dds/participant/ParticipantTests.cpp (hunk 14 above)
Comment (imontesino) [reply test/unittest/dds/participant/ParticipantTests.cpp:598]: It is a mistake, it has now been reset.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2920/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/203/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2670/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2931/)

**Pull request merged**